File: model/modules/cCycleConsistency/cCycleConsistency_simple/cCycleConsistency_simple.m
function [f, fe, fx, s, d, p] = cCycleConsistency_simple(f, fe, fx, s, d, p, info, tix)
    % +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
    % check consistency in cCycle matrix: cAlloc, cFlow
    %
    % Inputs:
    %	- s.cd.cAlloc:            carbon allocation matrix
    %   - flow_matrix:            carbon flow matrix
    %
    % Outputs:
    %   -
    %
    % Modifies:
    % 	-
    %
    % References:
    %	-
    %
    % Created by:
    %   - Simon Besnard (sbesnard)
    %
    % Versions:
    %   - 1.0 on 12.03.2020
    % +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

    % check allocation
    tmp0 = s.cd.cAlloc(:); %sujan
    tmp1 = sum(s.cd.cAlloc, 2);

    if any(tmp0 > 1) || any(tmp0 < 0)
        error('SINDBAD TEM: cAlloc lt 0 or gt 1')
    end

    if any(abs(sum(tmp1, 2) - 1) > 1E-6)
        error('SINDBAD TEM: sum(cAlloc) ne1')
    end

    % Check carbon flow matrix
    % the sum of A per column below the diagonals is always < 1
    % sujan: 22/03/2021: the flow_matrix reshape here is extremely slow..Also, it will not work when there is more than 1 pixel.
    for pix = 1:info.tem.helpers.sizes.nPix
        flow_matrix = squeeze(s.cd.p_cFlowAct_A(pix));
        flagUp = triu(ones(size(flow_matrix)), 1);
        flagLo = tril(ones(size(flow_matrix)), -1);
        % of diagonal values of 0 must be between 0 and 1
        anyBad = any(flow_matrix .* (flagLo + flagUp) < 0);

        if anyBad
            error('negative values in the p_cFlowAct_A matrix!')
        end

        anyBad = any(flow_matrix .* (flagLo + flagUp) > 1 + 1E-6);

        if anyBad
            error('values in the p_cFlowAct_A matrix greater than 1!')
        end

        % in the lower and upper part of the matrix A the sums have to be lower than 1
        anyBad = any(sum(flow_matrix .* flagLo, 1) > 1 + 1E-6);

        if anyBad
            error('sum of cols higher than one in lower in p_cFlowAct_A matrix')
        end

        anyBad = any(sum(flow_matrix .* flagUp, 1) > 1 + 1E-6);

        if anyBad
            error('sum of cols higher than one in upper in p_cFlowAct_A matrix')
        end

    end

    % flow_matrix = reshape(s.cd.p_cFlowAct_A, info.tem.model.variables.states.c.nZix.cEco, info.tem.model.variables.states.c.nZix.cEco);
    % flagUp = triu(ones(size(flow_matrix)),1);
    % flagLo     = tril(ones(size(flow_matrix)),-1);
    % % of diagonal values of 0 must be between 0 and 1
    % anyBad     = any(flow_matrix.*(flagLo+flagUp) < 0);
    % if anyBad
    %     error('negative values in the p_cFlowAct_A matrix!')
    % end
    % anyBad     = any(flow_matrix.*(flagLo+flagUp) > 1 + 1E-6);
    % if anyBad
    %     error('values in the p_cFlowAct_A matrix greater than 1!')
    % end
    % % in the lower and upper part of the matrix A the sums have to be lower than 1
    % anyBad     = any(sum(flow_matrix.*flagLo,1)>1 + 1E-6);
    % if anyBad
    %     error('sum of cols higher than one in lower in p_cFlowAct_A matrix')
    % end
    % anyBad     = any(sum(flow_matrix.*flagUp,1)>1 + 1E-6);
    % if anyBad
    %     error('sum of cols higher than one in upper in p_cFlowAct_A matrix')
    % end

end

File: src/models/cCycleConsistency/cCycleConsistency_simple.jl
export cCycleConsistency_simple

struct cCycleConsistency_simple <: cCycleConsistency
end

function compute(o::cCycleConsistency_simple, forcing, land::NamedTuple, helpers::NamedTuple)

	## unpack land variables
	@unpack_land begin
		cAlloc ∈ land.states
		p_A ∈ land.cFlow
		(𝟘, 𝟙, tolerance) ∈ helpers.numbers
	end

	# check allocation
	if any(cAlloc .> 𝟙)
		@show cAlloc
		error("cAllocation is greater than 1. Cannot continue")
	end
	if any(cAlloc .< 𝟘)
		@show cAlloc
		error("cAllocation is negative. Cannot continue")
	end
	if !isapprox(sum(cAlloc), 𝟙; atol=tolerance)
		@show cAlloc, sum(cAlloc)
		error("cAllocation does not sum to 1. Cannot continue")
	end

	# Check carbon flow matrix
	# the sum of A per column below the diagonals is always < 1
	offDiagA = offDiag(p_A)
	offDiagU = offDiagUpper(p_A)
	offDiagL = offDiagLower(p_A)
	if any(offDiagA .< 𝟘)
		@show offDiagA, p_A
		error("negative values in flow matrix. Cannot continue")
	end
	if any(offDiagA .> 𝟙)
		@show offDiagA, p_A
		error("flow is greater than 1. Cannot continue")
	end

	# if !isapprox(sum(offDiagL), 𝟙; atol=tolerance)
	# 	@show sum(offDiagL), offDiagL, p_A
	# 	error("sum of cols greater than one in lower cFlow matrix")
	# end
	# if !isapprox(sum(offDiagU), 𝟙; atol=tolerance)
	# 	@show sum(offDiagU), offDiagU, p_A
	# 	error("sum of cols greater than one in upper cFlow matrix")
	# end
	return land
end

@doc """
check consistency in cCycle matrix: cAlloc; cFlow

---

# compute:
Consistency checks on the c allocation and transfers between pools using cCycleConsistency_simple

*Inputs*
 - flow_matrix: carbon flow matrix
 - land.states.cAlloc: carbon allocation matrix

*Outputs*

# precompute:
precompute/instantiate time-invariant variables for cCycleConsistency_simple


---

# Extended help

*References*

*Versions*
 - 1.0 on 12.03.2020  

*Created by:*
 - sbesnard
"""
cCycleConsistency_simple